### PR TITLE
Feature/cluster zones

### PR DESCRIPTION
**Pull request by julsemaan**
# Description
Adds the ability to have clusters that are spread across multiple L3 networks.

# NEW Package(s) required
fingerbank-collector-1.1.0

# Impacts
Pretty much all PF as I had to rework stuff in pfconfig

# Delete branch after merge
YES

# NEWS file entries
(REQUIRED, but may be optional...)
## New Features
* Added support for clusters that have servers in different layer 3 networks

### CODE REVIEW

**Comment by extrafu** [comment]
I suggest also:

1- some documentation that explains this feature
2- a procedure to take a server from a L2 cluster and moving it into a separate cluster zone

**Comment by nqb** [comment]
Agree with @extrafu

**Comment by julsemaan** [comment]
Documentation will be for 9.0 :)

Jokes aside, yes I'll attack the documentation when I get back from vacation after the rush of 8.2 is over

**Comment by extrafu** [comment]
@jrouzierinverse Make sure your review this ASAP to make some iterations before next week.

File: lib/pf/cluster.pm [hunk 1]
```diff
@@ -50,22 +50,29 @@ use Module::Pluggable
...
-tie @cluster_hosts, 'pfconfig::cached_array', 'resource::cluster_hosts';
-$cluster_enabled = sub {
-    my $cfg = Config::IniFiles->new( -file => $cluster_config_file );
-    return 0 unless($cfg);
-    my $mgmt_ip = $cfg->val('CLUSTER', 'management_ip');
-    defined($mgmt_ip) && valid_ip($mgmt_ip) ? 1 : 0 ;
-}->();
+@EXPORT = qw(%ConfigCluster @cluster_servers @cluster_hosts $cluster_enabled $host_id $CLUSTER $cluster_name);
+
+use Carp; carp "here";
```
Comment (jrouzierinverse) [inline lib/pf/cluster.pm]: The ``` carp "here"; ``` is causes a lot of tests to fail.

File: lib/pfconfig/namespaces/resource/clusters_hostname_map.pm [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+=cut
+
+use strict;
+use warnings;
+
+use base 'pfconfig::namespaces::resource';
+use pfconfig::namespaces::config::Cluster;
+
+sub init {
```
Comment (jrouzierinverse) [inline lib/pfconfig/namespaces/resource/clusters_hostname_map.pm:27]: pod doc

File: lib/pfconfig/namespaces/resource/clusters_hostname_map.pm [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    my ($self) = @_;
+
+    $self->{cluster_resource} = pfconfig::namespaces::config::Cluster->new($self->{cache});
+}
+
+sub build {
```
Comment (jrouzierinverse) [inline lib/pfconfig/namespaces/resource/clusters_hostname_map.pm:39]: pod doc

File: lib/pfconfig/namespaces/resource/all_cluster_hosts.pm [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+pfconfig::namespaces::resource::all_cluster_hosts
+=cut
+
+use strict;
+use warnings;
+
+use base 'pfconfig::namespaces::resource';
+use pfconfig::namespaces::config::Cluster;
+
+sub init {
```
Comment (jrouzierinverse) [inline lib/pfconfig/namespaces/resource/all_cluster_hosts.pm:23]: pod doc

File: lib/pfconfig/namespaces/resource/all_cluster_hosts.pm [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    my ($self, $cluster_name) = @_;
+
+    $self->{cluster_name} = $cluster_name || "DEFAULT";
+    $self->{cluster_resource} = pfconfig::namespaces::config::Cluster->new($self->{cache});
+}
+
+sub build {
```
Comment (jrouzierinverse) [inline lib/pfconfig/namespaces/resource/all_cluster_hosts.pm:36]: pod doc

File: lib/pfconfig/namespaces/resource/all_cluster_servers.pm [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+=head1 NAME
+
+pfconfig::namespaces::resource::all_cluster_servers
+
+=cut
+
+=head1 DESCRIPTION
+
+pfconfig::namespaces::resource::all_cluster_servers
```
Comment (jrouzierinverse) [inline lib/pfconfig/namespaces/resource/all_cluster_servers.pm:11]: Better description

File: lib/pfconfig/namespaces/resource/all_cluster_servers.pm [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+=cut
+
+use strict;
+use warnings;
+
+use pfconfig::namespaces::config::Cluster;
+
+use base 'pfconfig::namespaces::resource';
+
+sub init {
```
Comment (jrouzierinverse) [inline lib/pfconfig/namespaces/resource/all_cluster_servers.pm:22]: pod doc

File: lib/pfconfig/namespaces/resource/all_cluster_servers.pm [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    my ($self, $cluster_name) = @_;
+
+    $self->{cluster_name} = $cluster_name || "DEFAULT";
+    $self->{cluster_resource} = pfconfig::namespaces::config::Cluster->new($self->{cache});
+}
+
+sub build {
```
Comment (jrouzierinverse) [inline lib/pfconfig/namespaces/resource/all_cluster_servers.pm:29]: pod doc

File: lib/pfconfig/namespaces/resource/all_cluster_hosts.pm [hunk 9]
```diff
@@ -0,0 +1,64 @@
+package pfconfig::namespaces::resource::all_cluster_hosts;
+
+=head1 NAME
+pfconfig::namespaces::resource::all_cluster_hosts
+=cut
+
+=head1 DESCRIPTION
+pfconfig::namespaces::resource::all_cluster_hosts
```
Comment (jrouzierinverse) [inline lib/pfconfig/namespaces/resource/all_cluster_hosts.pm:8]: Better description

File: lib/pfconfig/namespaces/resource/clusters_hostname_map.pm [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+=head1 NAME
+
+pfconfig::namespaces::resource::clusters_hostname_map
+
+=cut
+
+=head1 DESCRIPTION
+
+pfconfig::namespaces::resource::clusters_hostname_map
```
Comment (jrouzierinverse) [inline lib/pfconfig/namespaces/resource/clusters_hostname_map.pm:11]: better description

File: sbin/pf-mariadb [hunk 11]
```diff
@@ -161,12 +161,12 @@ Whether or not there is a ping quorum (i.e. servers are booted and online)
...
 sub ping_quorum {
     my $alive = 0;
-    foreach my $server (pf::cluster::enabled_servers()) {
+    foreach my $server (pf::cluster::mysql_servers()) {
         if(ping($server->{management_ip})) {
             $alive ++;
         }
     }
-    return ($alive > (scalar(pf::cluster::enabled_servers()) / 2));
+    return ($alive > (scalar(pf::cluster::db_enabled_hosts()) / 2));
```
Comment (jrouzierinverse) [inline sbin/pf-mariadb]: I understand that these are the same underneath.
However It is confusing to see two different functions to get the same array.
Suggestion to make thing clearer
```perl
my @servers = pf::cluster::mysql_servers();
foreach my $server (@servers) {
        if(ping($server->{management_ip})) {
            $alive ++;
        }
}
return ($alive > (scalar(@servers) / 2));
````

File: lib/pf/cluster.pm [hunk 12]
```diff
@@ -98,6 +105,56 @@ sub enabled_hosts {
...
 }
 
+=head2 db_enabled_servers
+
+Returns the @db_cluster_servers list without the servers that are disabled on this host
+
+=cut
+
+sub db_enabled_servers {
+    return map { (-f node_disabled_file($_->{host})) ? () : $_ } @db_cluster_servers;
```
Comment (jrouzierinverse) [inline lib/pf/cluster.pm]: grep would be more appropriate.

```perl
     return grep { -f node_disabled_file($_->{host}) }  @db_cluster_servers;
```

File: lib/pf/cluster.pm [hunk 13]
```diff
@@ -98,6 +105,56 @@ sub enabled_hosts {
...
+}
+
+=head2 db_enabled_hosts
+
+Returns the @db_cluster_hosts list without the servers that are disabled on this host
+
+=cut
+
+sub db_enabled_hosts {
+    return map { (-f node_disabled_file($_)) ? () : $_ } @db_cluster_hosts;
```
Comment (jrouzierinverse) [inline lib/pf/cluster.pm]: grep would be more appropriate.

File: lib/pf/cluster.pm [hunk 14]
```diff
@@ -98,6 +105,56 @@ sub enabled_hosts {
...
+}
+
+=head2 config_enabled_hosts
+
+Returns the @config_cluster_hosts list without the servers that are disabled on this host
+
+=cut
+
+sub config_enabled_hosts {
+    return map { (-f node_disabled_file($_)) ? () : $_ } @config_cluster_hosts;
```
Comment (jrouzierinverse) [inline lib/pf/cluster.pm]: grep would be more appropriate.

File: lib/pf/cluster.pm [hunk 15]
```diff
@@ -98,6 +105,56 @@ sub enabled_hosts {
...
+}
+
+=head2 config_enabled_servers
+
+Returns the @config_cluster_servers list without the servers that are disabled on this host
+
+=cut
+
+sub config_enabled_servers {
+    return map { (-f node_disabled_file($_->{host})) ? () : $_ } @config_cluster_servers;
```
Comment (jrouzierinverse) [inline lib/pf/cluster.pm]: grep would be more appropriate.

File: lib/pf/cluster.pm [hunk 16]
```diff
@@ -195,13 +252,13 @@ Get the list of the MySQL servers ordered by priority
 =cut
 
 sub mysql_servers {
-    if(scalar(enabled_servers()) >= 1){
+    if(scalar(db_enabled_hosts()) >= 1){
```
Comment (jrouzierinverse) [inline lib/pf/cluster.pm]: just return

```perl
  return reverse db_enabled_hosts();
```
Simplifies the logic and has the same results.

**Review by jrouzierinverse — CHANGES_REQUESTED** [review]

File: sbin/pf-mariadb (hunk 11 above)
Comment (jrouzierinverse) [reply sbin/pf-mariadb]: Missed this?

**Comment by jrouzierinverse** [comment]
Is there still more work to be done?

**Comment by julsemaan** [comment]
Some adjustments based on my tests and I still need to complete the adjustments you put comments

**Comment by julsemaan** [comment]
Its ready for another review

**Comment by jrouzierinverse** [comment]
Were the dependencies updated for fingerbank-collector

File: lib/pf/api.pm [hunk 17]
```diff
@@ -656,7 +657,7 @@ sub notify_configfile_changed : Public {
     };
     pfconfig::util::fetch_decode_socket(encode_json($payload));
 
-    my $master_server = $ConfigCluster{$postdata{server}};
+    my $master_server = firstval { $_->{host} eq $postdata{server} } pf::cluster::config_enabled_servers;
```
Comment (jrouzierinverse) [inline lib/pf/api.pm]: Minor stuff make this a function.

Comment (jrouzierinverse) [reply lib/pf/api.pm]: Then I will run a smoke test then merge.

**Pull request merged**